### PR TITLE
Feature/host request

### LINKED ISSUE #23: Support multiple KakapoServer
As for now we need to use class methods

### LINKED ISSUE #24: Support host for request stub
- [ ] Implement #23 

needs multiple kakapo server support #23

### LINKED ISSUE #34: Unregister routes
Say you register:

``` swift
KakapoServer.get("/user/:id")...
```

We may want to support unregistering routes, this might be automatically done when Server/Routers are deallocated but this can only happen when #23 is done.

@zzarcon alignment needed ?
- [x] Implement #23 
- [x] Test unregister routes
- [x] Test replacing routes
- [x] Test dealloc unregister all routes

### LINKED ISSUE #43: Document KakapoServer and KakapoDB
#23
#24

### PR DESCRIPTION
**Pull request by joanromano**
Closes #23 
Closes #24 
Closes #43 
Closes #34 

To discuss:
- [x] Whether new `decomposeURL` is flexible enough
- [x] Whether registration of routers with scheme (e.g `http`) should fail afterwards and so registration should always be without it, only host (`www.foo.com`) -> See tests for more info
- [x] Whether we want to unregister by host string instead of handling the router instance in the unregister method of `KakapoServer` (currently by host string) -> See tests for more info
- [x] Whether we shouldn't use `KakapoServer` at all to register Router but instead always the `Router` class.

### CODE REVIEW

**Comment by MP0w** [comment]
And documentation

**Comment by joanromano** [comment]
@MP0w I was going to do that in a different PR in order to handle #43, what do you think?

File: Source/KakapoServer.swift [hunk 1]
```diff
@@ -33,61 +33,54 @@ public struct Response: CustomSerializable {
...
 
-public class KakapoServer: NSURLProtocol {
-
-    private typealias Route = (method: HTTPMethod, handler: RouteHandler)
+public class Router {
     
     private enum HTTPMethod: String {
         case GET, POST, PUT, DELETE
     }
+    private typealias Route = (method: HTTPMethod, handler: RouteHandler)
```
Comment (MP0w) [inline Source/KakapoServer.swift]: newline

File: Source/KakapoServer.swift [hunk 2]
```diff
@@ -33,61 +33,54 @@ public struct Response: CustomSerializable {
...
-    public class func disable() {
-        routes = [:]
-        NSURLProtocol.unregisterClass(self)
-    }
-    
-    override public class func canInitWithRequest(request: NSURLRequest) -> Bool {
-        guard let requestString = request.URL?.absoluteString else { return false }
+    func canInitWithRequest(request: NSURLRequest) -> Bool {
+        guard let requestURL = request.URL,
+                  components = requestURL.componentsFromBaseURL(baseURL) else { return false }
```
Comment (MP0w) [inline Source/KakapoServer.swift]: why is this needed?

Comment (joanromano) [inline Source/KakapoServer.swift]: Remove baseURL from incoming request to use it when decomposing the URL

File: Source/KakapoDB.swift [hunk 3]
```diff
@@ -8,11 +8,17 @@
 
 import Foundation
 
+/**
+ A protocol to store objects into a store, providing an `id` and a generic `init` method for all objects that will be inserted
+ */
```
Comment (MP0w) [inline Source/KakapoDB.swift:15]: Maybe you should mention why it's needed (kind of type erasure)

File: Source/KakapoDB.swift [hunk 4]
```diff
@@ -63,6 +69,14 @@ public class KakapoDB {
         return object!
     }
     
+    /**
+     Inserts Storable objects based on their default initializer
```
Comment (MP0w) [inline Source/KakapoDB.swift]: Create storable objects and insert them in the database

File: Source/KakapoDB.swift [hunk 5]
```diff
@@ -63,6 +69,14 @@ public class KakapoDB {
...
+     
+     - parameter (unamed): The Storable Type to be created
+     - parameter number: The number of elements to create, defaults to 1
+     
+     - returns: An array containing the new inserted Storable objects
+     */
     public func create<T: Storable>(_: T.Type, number: Int = 1) -> [T] {
         let ids = barrierSync {
```
Comment (MP0w) [inline Source/KakapoDB.swift:100]: I'm wondering why we don't call `insert` for create, to avoid duplicate/similar code

File: Source/KakapoServer.swift [hunk 6]
```diff
@@ -8,21 +8,42 @@
 
 import Foundation
 
+/**
+ A RouteHandler used when registering different HTTP methods. This can return any Serializable object, though you would normally want to use the built-in Response object
+ */
```
Comment (MP0w) [inline Source/KakapoServer.swift]: Not sure what you mean... Also mention that returning a `Response` you can provide status code and headers otherwise will use 200 by default

File: Source/KakapoServer.swift [hunk 7]
```diff
@@ -8,21 +8,42 @@
...
+    
+    /// An optional request body
     public let HTTPBody: NSData?
+    
+    /// An optional dictionary holding the request header fields
     public let HTTPHeaders: [String: String]?
 }
 
+/**
+ A Response struct which can be used in `RouteHandlers` to provide valid responses.
```
Comment (MP0w) [inline Source/KakapoServer.swift]: Please explain why you would need it (provide response code and headers)

File: KakapoExample/KakapoTests/checkUrlTest.swift [hunk 8]
```diff
@@ -15,24 +15,26 @@ class checkUrlTest: QuickSpec {
     override func spec() {
         describe("#checkUrl") {
             it("should return nil if the requested url doesn't match the declared one") {
-                expect(parseUrl("/users/:id", requestURL: "/comments/1")).to(beNil())
```
Comment (MP0w) [inline KakapoExample/KakapoTests/checkUrlTest.swift:18]: This file should be called `DecomposeURLTests`? describe could be more describing and these tests are actually testing multiple behavior in single tests (we may rewrite it as part of another PR if you prefer but create an issue then)

Comment (MP0w) [inline KakapoExample/KakapoTests/checkUrlTest.swift:18]: And I guess decompose could need more tests (will skip the review of this tests for now)

File: Source/KakapoServer.swift [hunk 9]
```diff
@@ -1,129 +1,69 @@
...
-    private enum HTTPMethod: String {
-        case GET, POST, PUT, DELETE
+    /**
+     Register and return a new Router in the Server
+     
+     - parameter baseURL: The base URL that this Router will use
+     
+     - returns: An new initialized Router. Note that two Router objects can hold the same baseURL.
+     */
+    class func register(baseURL: String) -> Router {
```
Comment (MP0w) [inline Source/KakapoServer.swift:25]: Not thread safe but I guess we don't need it

File: Source/KakapoServer.swift [hunk 10]
```diff
@@ -1,129 +1,69 @@
...
-    public static func del(urlString: String, handler: RouteHandler) {
-        KakapoServer.routes[urlString] = (.DELETE, handler)
-    }
-    
-    public static func put(urlString: String, handler: RouteHandler) {
-        KakapoServer.routes[urlString] = (.PUT, handler)
+        KakapoServer.routers.filter { URL.absoluteString.containsString($0.baseURL) && $0.canInitWithRequest(request) }.first?.startLoading(self)
     }
     
+    override func stopLoading() {}
```
Comment (MP0w) [inline Source/KakapoServer.swift:68]: if/when we introduce latency we might have to implement this

File: Source/KakapoDB.swift [hunk 11]
```diff
@@ -8,11 +8,21 @@
...
+ */
 public protocol _Storable {
     var id: Int { get }
     init(id: Int, db: KakapoDB)
 }
 
+/**
+ A protocol that supports both `_Storable` and `Equatable` objects, handling an `id` and a generic `init` method, as well as value equality. 
+ 
+ This is the public method which will be used in KakapoDB
```
Comment (MP0w) [inline Source/KakapoDB.swift]: the public protocol that will be required by `KakapoDB`

File: Source/KakapoDB.swift [hunk 12]
```diff
@@ -8,11 +8,21 @@
...
+ */
```
Comment (MP0w) [inline Source/KakapoDB.swift:25]: KakaDB class is not documented

File: Source/KakapoDB.swift [hunk 13]
```diff
@@ -120,16 +159,39 @@ public class KakapoDB {
...
+     - parameter includeElement: The predicate to satisfy the filtering
+     
+     - returns: An array containing the filtered Storable objects
+     */
     public func filter<T: Storable>(_: T.Type, includeElement: (T) -> Bool) -> [T] {
         return findAll(T).filter(includeElement)
     }
     
+    /**
+     Find all the objects in the store by a given id
```
Comment (MP0w) [inline Source/KakapoDB.swift]: "Find the" is only one or zero

File: Source/Router.swift [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+    
+    init(baseURL: String) {
+        self.baseURL = baseURL
+    }
+    
+    func canInitWithRequest(request: NSURLRequest) -> Bool {
+        guard let requestURL = request.URL,
+                  components = requestURL.componentsFromBaseURL(baseURL) else { return false }
+        
+        for (key, route) in routes where route.method.rawValue == request.HTTPMethod {
```
Comment (MP0w) [inline Source/Router.swift:115]: 😍

File: Source/URLDecomposer.swift [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+import Foundation
+
+/**
+ A tuple holding components and query parameters, check `decomposeURL` for more details
+ */
+public typealias URLInfo = (components: [String : String], queryParameters: [NSURLQueryItem])
+
+/**
+ Checks and parses if a given `handlerPath` representation matches a `requestURLComponents`. Examples:
```
Comment (MP0w) [inline Source/URLDecomposer.swift]: Parses and checks

File: Source/URLDecomposer.swift [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+ `/users/:id/comments` with `/users/1/comments` produces `components: ["id" : "1"]`
+ `/users/:id/comments/:comment_id` with `/users/1/comments/2?page=2&author=hector` produces `components: ["id": "1", "comment_id": "2"]` and `queryParameters: ["page": "2", "author": "hector"]`
+ 
+ - parameter handlerPath: the URL handler with dynamic paths
+ - parameter requestURLComponents: the components of the actual URL
+ 
+ - returns: a URL info object containing `components` and `queryParamaters`
+ */
+func decomposeURL(handlerPath: String, requestURLComponents: NSURLComponents) -> URLInfo? {
+    var components: [String : String] = [:]
```
Comment (MP0w) [inline Source/URLDecomposer.swift:29]: a bit difficult to understand, would you mind adding comments with the expected result after the expressions? like `let handlerSplittedPaths.... // ["/user/", ":id/comment", "commentId"]`

Comment (MP0w) [inline Source/URLDecomposer.swift:29]: I wonder if it can be simplified

Comment (joanromano) [inline Source/URLDecomposer.swift:29]: I'd leave that for a different PR with an issue for discussing it.

Comment (MP0w) [inline Source/URLDecomposer.swift:29]: yes okay, I just can't understand how it works, let's leave it

Comment (joanromano) [inline Source/URLDecomposer.swift:29]: totally agree, it's not the easiest solution. I'll create an issue

File: Source/URLDecomposer.swift (hunk 15 above)
Comment (MP0w) [inline Source/URLDecomposer.swift]: just because "Parse if a given" sound really wrong

Comment (MP0w) [inline Source/URLDecomposer.swift]: oh no maybe I'm reading it wrong, is like:

```
if a given handlerPath representation matches....
    check and parses
```

?

File: KakapoExample/KakapoTests/RouterTests.swift [hunk 17]
```diff
@@ -0,0 +1,720 @@
...
+                expect(info?.components).toEventually(equal(["id" : "1"]))
+                expect(info?.queryParameters).toEventually(equal([]))
+                expect(responseURL?.absoluteString).toEventually(equal("http://www.test.com/users/1"))
+                
+                expect(secondInfo?.components).toEventually(equal(["user" : "24"]))
+                expect(secondInfo?.queryParameters).toEventually(equal([]))
+                expect(secondResponseURL?.absoluteString).toEventually(equal("http://www.host2.com/messages/24"))
+            }
+            
+            it ("Should not handle invalid registrations on routers (leading and trailing slashes should match the absolute request)") {
```
Comment (MP0w) [inline KakapoExample/KakapoTests/RouterTests.swift:432]: O RLY? so if you miss the trailing won't work? BTW these tests should be in the URL matcher not in the server.

Server: test if matched urls callback
Decomposer: test the matcher

Comment (MP0w) [inline KakapoExample/KakapoTests/RouterTests.swift:432]: I thought that was easily possible to support with NSURL but we can see that later

File: KakapoExample/KakapoTests/URLDecomposerTests.swift [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+            }
+            
+            it("it should match a handler path and request path with same wilcard combinations") {
+                expect(decomposeURL("/users/:id/comments", requestURLComponents: NSURLComponents(string: "/users/1/comments")!)!.components) == ["id" : "1"]
+            }
+            
+            it("it should match a handler path and request path with same wilcard combinations 2") {
+                expect(decomposeURL("/users/:user_id/comments/:comment_id", requestURLComponents: NSURLComponents(string: "/users/1/comments/2")!)!.components) == ["user_id" : "1", "comment_id": "2"]
+            }
+        }
```
Comment (MP0w) [inline KakapoExample/KakapoTests/URLDecomposerTests.swift:54]: it should match a wildcard when preceded by multiple components
it should match a wildcard when there are no components (":id" or ":id/:commentId")

File: KakapoExample/KakapoTests/URLDecomposerTests.swift [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+        
+        describe("matching with query parameters") {
+            it("it should match a handler path with no wilcard with the same request path with query parameters, providing the proper components") {
```
Comment (MP0w) [inline KakapoExample/KakapoTests/URLDecomposerTests.swift:57]: it should match query parameters when proceeded by a trailing slash (Users/?id=2) I guess

**Comment by MP0w** [comment]
LGTM 👍
